app/utils/events_utils.py

```diff
@@ -0,0 +1,231 @@
+            categories = ["events"]
+
+        if start_date > end_date:
+            raise ValueError("Start date cannot be after end date")
+
+        # Join categories
+        categories_str = ", ".join(categories)
+        logger.debug(f"Formatted categories: {categories_str}")
+
+        # Read template from file
+        try:
+            template = read_prompt_from_file("app/prompts/events_query_template.txt")
+            logger.debug("Successfully loaded query template")
+        except FileNotFoundError as e:
+            logger.error(f"Template file not found: {e}")
+            raise
+        except Exception as e:
+            logger.error(f"Failed to read template file: {e}")
+            raise
+
+        # Format the template with the provided parameters
+        try:
+            formatted_query = template.format(
+                categories=categories_str,
+                city=city.strip(),
+                start_date=start_date.strftime("%B %d"),
+                end_date=end_date.strftime("%B %d"),
+            )
+            logger.info("Successfully built query string")
+            return formatted_query
+
+        except KeyError as e:
+            logger.error(f"Template formatting failed - missing placeholder: {e}")
+            raise ValueError(f"Template formatting error: {e}")
+        except Exception as e:
+            logger.error(f"Unexpected error during template formatting: {e}")
+            raise
+
+    except Exception as e:
+        logger.error(f"Failed to build query: {e}")
+        raise
+
+
+def parse_event_query(user_input: str, structured_llm) -> EventQuery | None:
+    """
+    Try to parse an EventQuery from user input.
+    If dates are missing, enrich with pre-processed hints.
+    """
+    try:
+        hints = preprocess_dates(user_input)
+
+        # Pass today's date as contextual hint to the parser to reduce wrong-year outputs
+        current_date = date.today()
+        augmented_input = (
+            f"{user_input}\n\n"
+            f"Note: The current year is {current_date.year}. Make sure the start date corresponds to the current year.\n"
+            f"If the user doesn't state a specific date, set start_date to today's date {current_date} and set end_date accordingly (use duration if provided, otherwise equal to start_date)."
+        )
+        query = structured_llm.invoke(augmented_input)
+
+        # Enrich dates if missing
+        if not query.start_date and hints.get("start_date"):
+            query.start_date = hints["start_date"]
+        if not query.end_date and hints.get("end_date"):
+            query.end_date = hints["end_date"]
+
+        return query
+
+    except ValidationError as e:
+        logging.warning(f"Invalid EventQuery schema, skipping. Error: {e}")
+        return None
+    except Exception as e:
+        logging.error(f"Unexpected error in EventQuery parsing: {e}")
+        return None
+
+
+def extract_events(raw: str) -> List[Dict[str, Any]]:
```

Comment:
The extraction does not handle single event dictionaries, only lists. Add logic to wrap non-list dicts into a list.
```
if parsed:
    if isinstance(parsed, list):
        return parsed
    elif isinstance(parsed, dict):
        return [parsed]
```